google/cloud/aiplatform/base.py

```diff
@@ -16,13 +16,196 @@
 #
 
 import abc
-from typing import Optional
+from concurrent import futures
+import functools
+import inspect
+import threading
+from typing import Any, Callable, Dict, Optional, Sequence, Union
 
 from google.auth import credentials as auth_credentials
 from google.cloud.aiplatform import utils
 from google.cloud.aiplatform import initializer
 
 
+class FutureManager(metaclass=abc.ABCMeta):
+    """Tracks concurrent futures against this object."""
+
+    def __init__(self):
+        self.__latest_future_lock = threading.Lock()
+
+        # Always points to the latest future. All submitted futures will always
+        # form a dependency on the latest future.
```

Comment:
Does this mean that if future1 and future2 are submitted in that order, then self.__latest_future will point to future1, while future2 will be remembered as a dependency of future1.  The moment future1 is resolved, future2 becomes self.__latest_future?

Reply:
Related: do we see the need for the user to be able to inspect the completed and yet-to-complete futures for debugging?

Reply:
> The moment future1 is resolved, future2 becomes self.__latest_future?

`future2` will become the latest future immediately so any down stream call against this noun can form a dependency on future2. 

I'll change this comment to say:

```python
# Always points to the latest future. All submitted futures will always
# form a dependency on the latest future and then be set as the latest future.
```

Reply:
Depending on the user's perspective, the term "latest" might be interpreted as future2 in the example above even before future1 has been completed, since future2 was submitted the latest.  But i don't think this is a serious problem.